### ISSUE TITLE
TPublisher is not in the include folder

### ISSUE THREAD

**Issue by proostas** [treefrogframework/treefrog-framework#286]
Hello folks,

Is there a reason why tpublisher.h is not copied in /usr/include/treefrog?

Can I use TPublisher in a model or controller? It seems to be the only way to push WS message from there.

**Comment by treefrogframework** [comment]
No.

See https://www.treefrogframework.org/en/user-guide/websocket/

**Comment by proostas** [comment]
Thank you for you reply. 

I have read the documentation regarding TWebSocketEndpoint and currently understand how it works for external clients (the ones that establish connection to the end point from outside of the server).

But my problem is about sending WebSocket messages **internally** from a TActionController derivative. Currently I implemented this in a way I don't particularly like:
`
class T_HELPER_EXPORT EventNotifier : public TApplicationScheduler
{
public:
	EventNotifier();
public:
	void notify(const QJsonObject &json) {
              publish("new_event", QString(QJsonDocument(json).toJson()));
        };
};
`

`
class T_CONTROLLER_EXPORT ApplicationController : public TActionController
{
	Q_OBJECT
...
public slots:
       void staticInitialize() {
           s_eventNotifier = new EventNotifier;
       };
public:
       static EventNotifier *s_eventNotifier;
...
};
`

`
class T_CONTROLLER_EXPORT EventController : public ApplicationController
{
	Q_OBJECT
public:
	EventController() : ApplicationController() { }

public slots:
	void push() {
              ...
              QJsonObject ws_json {
		{"event", event.toJsonObject()},
		{"blob", blob.toJsonObject()},
		{"line", line.toJsonObject()},
		{"peer", Device::get(0).toJsonObject()}
	    };
	    s_eventNotifier->notify(ws_json);

	    renderSuccess(QJsonObject {{"id", event.id()}});
        };
};
`
This works. Sort of. But I am looking for a more elegant way of accomplishing this. I found following methods in a TActionController:

`
// For WebSocket
    void sendTextToWebSocket(int sid, const QString &text);
    void sendBinaryToWebSocket(int sid, const QByteArray &binary);
    void closeWebSokcet(int sid, int closeCode = Tf::NormalClosure);
`

... but I don't understand how I can use them to send message to a particular **topic**.

Also I have noticed that TScheduler internally uses ` TPublisher::instance()->publish(topic, text, nullptr);` to send WS messages so I don't get why I need to create a static helper since TPublisher::instance() is already static.

I am looking forward to your advice since the documentation does not cover this internal WebSocket message exchange and I have to study the source code (which is not a problem for me) to understand ways to accomplish this.
